parser.go

```diff
@@ -7,30 +7,53 @@ package properties
 import (
 	"fmt"
 	"runtime"
+	"strings"
 )
 
 type parser struct {
 	lex *lexer
 }
 
-func parse(input string) (properties *Properties, err error) {
-	p := &parser{lex: lex(input)}
+func parse(input string, preserveFormatting bool) (properties *Properties, err error) {
+	p := &parser{lex: lex(input, preserveFormatting)}
 	defer p.recover(&err)
 
 	properties = NewProperties()
+	properties.PreserveFormatting = preserveFormatting
 	key := ""
-	comments := []string{}
+	comments := []Comment{}
 
 	for {
 		token := p.expectOneOf(itemComment, itemKey, itemEOF)
 		switch token.typ {
 		case itemEOF:
+			if preserveFormatting && (len(comments) > 0 || token.val != "") {
+				// There are comments at the end of the input that are not tied to a particular key
+				// Save these off against a special empty key when preserving formatting
```

Comment:
Instead of a sentinel key I suggest using a separate field for heading and trailing comments, e.g. `trailComment comment` in the `Properties`.

Also, please revert the logic of the `if` statement to avoid the indent.

```go
if !preserveFormatting || len(comments) == 0 {
    goto done
}
if token.val == "" {
    p.trailingComments = comments
    goto done
}
...
goto done
```

Reply:
Done